synapse/state/__init__.py

```diff
@@ -140,34 +172,31 @@ async def get_current_state(
         state_map = await self.store.get_events(
             list(state.values()), get_prev_content=False
         )
-        state = {
+        return {
             key: state_map[e_id] for key, e_id in state.items() if e_id in state_map
         }
 
-        return state
-
-    async def get_current_state_ids(self, room_id, latest_event_ids=None):
+    async def get_current_state_ids(
+        self, room_id: str, latest_event_ids: Optional[Iterable[str]] = None
+    ) -> StateMap[str]:
         """Get the current state, or the state at a set of events, for a room
 
         Args:
-            room_id (str):
-
-            latest_event_ids (iterable[str]|None): if given, the forward
-                extremities to resolve. If None, we look them up from the
-                database (via a cache)
+            room_id:
+            latest_event_ids: if given, the forward extremities to resolve. If
+                None, we look them up from the database (via a cache).
 
         Returns:
-            Deferred[dict[(str, str), str)]]: the state dict, mapping from
-                (event_type, state_key) -> event_id
+            the state dict, mapping from (event_type, state_key) -> event_id
         """
         if not latest_event_ids:
             latest_event_ids = await self.store.get_latest_event_ids_in_room(room_id)
+        assert latest_event_ids is not None
 
         logger.debug("calling resolve_state_groups from get_current_state_ids")
         ret = await self.resolve_state_groups_for_events(room_id, latest_event_ids)
-        state = ret.state
-
-        return state
+        # mypy doesn't like that this returns a frozendict instead of a dict.
+        return ret.state  # type: ignore
```

Comment:
A better solution to this might be making `StateMap` a `Mapping` instead of a `Dict` or something? But I'm not 100% sure if that makes sense / is possible.

Reply:
It seems this can pretty easily be changed into a `MutableMapping` (not a `Mapping`), but that still isn't compatible with `frozendict` (since that isn't mutable).

Reply:
Thinking about this a bit more I think I'm more inclined to merge this as-is and come back to this, unless you see a straightforward way to improve this?

Reply:
I can't think of one, though can you use `cast(...)` instead of ignore?

Reply:
Though, can you copy your `frozendict` typing and make it a protocol?

Reply:
Adding a `dict()` call around the two locations fixed in (see 6f5515b462a41b664e15c458f9e8b8bcbcd38f03) although I'm not sure it's desirable.

A `cast` call might work, but is kind of lying to mypy?

I'm not sure how making a protocol would help?

Reply:
Well we're already lying to mypy by doing `#type: ignore`. The only real difference between that and cast is that cast is being explicit about what we're doing, rather than ignoring all errors on the line.

Wouldn't a protocol that had the common functions between frozendict and dict help?

Reply:
@erikjohnston So I ended up wrapping the calls in `dict()`, not using ignore.

I guess I need to read up more about what a protocol is to know if it would help here. I'll look into it!